app/code/Magento/CatalogInventory/Model/StockStateProvider.php

```diff
@@ -106,6 +106,10 @@ public function checkQuoteItemQty(StockItemInterface $stockItem, $qty, $summaryQ
     {
         $result = $this->objectFactory->create();
         $result->setHasError(false);
+        
+        $product = $this->productFactory->create();
+        $product->load($stockItem->getProductId());
+        $stockItem->setProduct($product);
```

Comment:
as your method accepts StockItemInterface as parameter, you should rely just on the contract provided by this interface https://github.com/magento/magento2/blob/develop/app/code/Magento/CatalogInventory/Api/Data/StockItemInterface.php
setProduct does not belong to StockItemInterface.
And couldn't belong, because StockItemInterface represents a relation between Stock and Product. You can consider it as a link entity, which holds the information about "what product" "in which quantity" stored on "what stock"